src/index.ts

```diff
@@ -1,3 +1,182 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 'use-strict';
+
+import {
+  IInputAreaViewModel
+} from 'jupyter-js-input-area';
+
+import {
+  IOutputAreaViewModel
+} from 'jupyter-js-output-area';
+
+import {
+  IObservableList
+} from 'phosphor-observablelist';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+import './index.css';
+
+
+/**
+ * An enum which describes the type of cell.
+ */
+enum CellType {
+  /**
+   * The cell contains code input.
+   */
+  Code,
+
+  /**
+   * The cell contains markdown.
+   */
+  Markdown,
+
+  /**
+   * The cell contains raw text.
+   */
+  Raw
+}
+
+
+/**
+ * An enum which describes which interface property has changed.
+ */
+export
+enum CellProperty {
+  CellType,
+  Tags,
+  Deletable,
+  Mergable,
+  Splittable,
+  Rendered,
+  Marked,
+  Format
+}
+
+
+/**
+ * The arguments object emitted with the `stateChanged` signal.
+ */
+export
+interface ICellChangedArgs<T> {
+  which: CellProperty,
+  oldValue: T;
+  newValue: T;
+}
+
+
+/**
+ * The definition of a model object for a base cell.
+ */
+interface IBaseCellViewModel {
+
+  /**
+   * The type of cell.
+   */
+  cellType: CellType;
+
+  /**
+   * Tags applied to the cell.
+   */
+  tags?: IObservableList<string>;
+
+  /**
+   * Get namespaced metadata about the cell.
+   */
+  getMetadata(namespace: string) : IObservableMap<string, JSONData>;
```

Comment:
Again, this point is not about type-safety for me; it's about ease of implementing the interface and proper separation of concerns. The thing(s) that will be shared across plugin boundaries should be models, not view models (typically), and I can easily see this metadata living there.

Here's some discussion on modelling a JSON value type (if you were curious):
https://github.com/Microsoft/TypeScript/issues/1897